### PR TITLE
wEth Vault and SSLP wBTC-wETH strategy integration

### PR DESCRIPTION
**Pull request by nodimdim**

### CODE REVIEW

File: src/both/DeltaNeutralLpV3.sol [hunk 1]
```diff
@@ -163,6 +173,9 @@ contract DeltaNeutralLpV3 is AccessStrategy {
     /// @notice Gives ratio of vault asset to borrow asset, e.g. WMATIC/USD (assuming usdc = usd)
     AggregatorV3Interface immutable borrowAssetFeed;
 
+    address public constant WETH = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
```
Comment (larryob) [inline src/both/DeltaNeutralLpV3.sol:176]: No need for this since we already have `borrowAsset`

File: src/both/DeltaNeutralLpV3.sol [hunk 2]
```diff
@@ -63,24 +63,34 @@ contract DeltaNeutralLpV3 is AccessStrategy {
...
     }
 
     /// @notice Convert `borrowAsset` (e.g. MATIC) to `asset` (e.g. USDC)
-    function _borrowToAsset(uint256 amountB, uint256 clPrice) internal pure returns (uint256 assets) {
-        // The first divisition gets rid of the decimals of wmatic. The second converts dollars to usdc
-        // TODO: make this work for any set of decimals
-        assets = amountB.mulWadDown(clPrice) / 1e2;
+    function _borrowToAsset(uint256 amountB, uint256 clPrice) internal view returns (uint256 assets) {
+        // The first divisition gets rid of the decimals of `borrowAsset`. The second converts dollars to `asset`
+        if (decimalDiff < 0) {
```
Comment (larryob) [inline src/both/DeltaNeutralLpV3.sol:81]: `decimalDiff` is a `uint` and cannot be less than zero.

File: src/both/DeltaNeutralLpV3.sol [hunk 3]
```diff
@@ -163,6 +173,9 @@ contract DeltaNeutralLpV3 is AccessStrategy {
     /// @notice Gives ratio of vault asset to borrow asset, e.g. WMATIC/USD (assuming usdc = usd)
     AggregatorV3Interface immutable borrowAssetFeed;
 
+    address public constant WETH = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+    uint256 public immutable decimalDiff;
```
Comment (larryob) [inline src/both/DeltaNeutralLpV3.sol:177]: I think we can use a function like [`_giveDecimals`](https://github.com/AffineLabs/contracts/blob/4c00d3f672e748c4eb39083e4c794feb000e55b8/src/ethereum/ConvexStrategy.sol#LL174C5-L182C6) to handle decimal conversions.

File: src/test/DeltaNeutralLpV3.t.sol [hunk 4]
```diff
@@ -182,11 +190,29 @@ contract DeltaNeutralV3EthTest is DeltaNeutralV3Test {
...
         strategy = SslpV3.deployEth(vault);
     }
 }
+
+contract DeltaNeutralV3EthWethTest is DeltaNeutralV3Test {
+    function _selectFork() internal override {
+        vm.createSelectFork("ethereum", 16_394_906);
+    }
+
+    function _asset() internal pure override returns (address) {}
```
Comment (larryob) [inline src/test/DeltaNeutralLpV3.t.sol:207]: I'm pretty sure the asset should be WETH here.

File: src/test/DeltaNeutralLpV3.t.sol [hunk 5]
```diff
@@ -182,11 +190,29 @@ contract DeltaNeutralV3EthTest is DeltaNeutralV3Test {
...
+
+    function _setAsset() internal virtual override {}
```
Comment (larryob) [inline src/test/DeltaNeutralLpV3.t.sol:209]: `_setAsset` does nothing now, so the asset is just the [fake token](https://github.com/AffineLabs/contracts/blob/4c00d3f672e748c4eb39083e4c794feb000e55b8/src/test/Deploy.sol#L34)  deployed along with the L2Vault.

**Review by larryob — COMMENTED** [review]
We can make a few changes here.

File: src/test/DeltaNeutralLpV3.t.sol (hunk 5 above)
Comment (nodimdim) [reply src/test/DeltaNeutralLpV3.t.sol:209]: No, the vault is not `L2Vault` here. It is a `Vault` deployed by the `EthVaults.deployEthWeth()`  function. See `_deployVault` override function.

File: src/test/DeltaNeutralLpV3.t.sol (hunk 4 above)
Comment (nodimdim) [reply src/test/DeltaNeutralLpV3.t.sol:207]: Removed in new version

File: src/both/DeltaNeutralLpV3.sol (hunk 3 above)
Comment (nodimdim) [reply src/both/DeltaNeutralLpV3.sol:177]: The new version is cleaner.

File: src/both/DeltaNeutralLpV3.sol (hunk 2 above)
Comment (nodimdim) [reply src/both/DeltaNeutralLpV3.sol:81]: My bad, don't what I was thinking. New version does not have this issue.

File: src/both/DeltaNeutralLpV3.sol (hunk 1 above)
Comment (nodimdim) [reply src/both/DeltaNeutralLpV3.sol:176]: Removed in the new version.

File: src/test/DeltaNeutralLpV3.t.sol (hunk 5 above)
Comment (nodimdim) [reply src/test/DeltaNeutralLpV3.t.sol:209]: And this vault has real WETH token as asset. So no need to override. Also, FYI, overriding this new `Vault` asset fails. Something to look into if we have time. Can't find slot for `asset()`.

```
Running 1 test for src/test/DeltaNeutralLpV3.t.sol:DeltaNeutralV3EthWethTest
[FAIL. Reason: Setup failed: stdStorage find(StdStorage): Slot(s) not found.] setUp() (gas: 0)
Test result: FAILED. 0 passed; 1 failed; finished in 3.02s
```

File: src/both/DeltaNeutralLpV3.sol [hunk 6]
```diff
@@ -131,6 +144,36 @@ contract DeltaNeutralLpV3 is AccessStrategy {
     /// @dev Aave deposit receipt token.
     ERC20 public immutable aToken;
 
+    /*//////////////////////////////////////////////////////////////
+                          PRICE CONVERSION
+    //////////////////////////////////////////////////////////////*/
```
Comment (larryob) [inline src/both/DeltaNeutralLpV3.sol:149]: There's already an exchange rates section below. You can move this code there.

File: src/both/DeltaNeutralLpV3.sol [hunk 7]
```diff
@@ -131,6 +144,36 @@ contract DeltaNeutralLpV3 is AccessStrategy {
...
+
+    /**
+     * @notice abs(asset.decimals() - borrow.decimals() - borrowFeed.decimals()). This value has significance
+     * while converting asset amount to borrow amount and vice versa.
```
Comment (larryob) [inline src/both/DeltaNeutralLpV3.sol:153]: The second sentence can say `Used when converting between asset/borrow amounts`. You can probably make this a single line comment and use `///`.

File: src/both/DeltaNeutralLpV3.sol [hunk 8]
```diff
@@ -131,6 +144,36 @@ contract DeltaNeutralLpV3 is AccessStrategy {
...
+     */
+    uint256 public decimalAdjust;
+    /**
+     * @notice true if `asset.decimals() - borrow.decimals() - borrowFeed.decimals()` is positive. false
+     * otherwise.
+     */
```
Comment (larryob) [inline src/both/DeltaNeutralLpV3.sol:159]: This is actually true if the expression is >= 0. You can make this into a single line comment as well.

File: src/both/DeltaNeutralLpV3.sol [hunk 9]
```diff
@@ -131,6 +144,36 @@ contract DeltaNeutralLpV3 is AccessStrategy {
...
 
+    /*//////////////////////////////////////////////////////////////
+                          PRICE CONVERSION
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice abs(asset.decimals() - borrow.decimals() - borrowFeed.decimals()). This value has significance
+     * while converting asset amount to borrow amount and vice versa.
+     */
+    uint256 public decimalAdjust;
```
Comment (larryob) [inline src/both/DeltaNeutralLpV3.sol:155]: This should be immutable.

File: src/both/DeltaNeutralLpV3.sol [hunk 10]
```diff
@@ -131,6 +144,36 @@ contract DeltaNeutralLpV3 is AccessStrategy {
...
+    /**
+     * @notice true if `asset.decimals() - borrow.decimals() - borrowFeed.decimals()` is positive. false
+     * otherwise.
+     */
+    bool public decimalAdjustSign;
+
+    /*//////////////////////////////////////////////////////////////
+                          STRATEGY PARAMS
+    //////////////////////////////////////////////////////////////*/
```
Comment (larryob) [inline src/both/DeltaNeutralLpV3.sol:164]: This section and the one below should be moved either above or below the position management section. Right now the position management section has been split into two pieces. See the `PositionStart` event and `startPosition`  function.

File: src/both/DeltaNeutralLpV3.sol (hunk 6 above)
Comment (larryob) [reply src/both/DeltaNeutralLpV3.sol:149]: You probably want to move these right above `_borrowToAsset()`

File: src/test/DeltaNeutralLpV3.t.sol [hunk 11]
```diff
@@ -182,11 +196,27 @@ contract DeltaNeutralV3EthTest is DeltaNeutralV3Test {
...
     }
 
     function _deployStrategy() internal override {
         strategy = SslpV3.deployEth(vault);
     }
 }
+
+contract DeltaNeutralV3EthWethTest is DeltaNeutralV3Test {
+    function _selectFork() internal override {
+        vm.createSelectFork("ethereum", 16_394_906);
```
Comment (larryob) [inline src/test/DeltaNeutralLpV3.t.sol:210]: Use `forkEth()`

File: src/both/DeltaNeutralLpV3.sol [hunk 12]
```diff
@@ -471,15 +514,27 @@ contract DeltaNeutralLpV3 is AccessStrategy {
...
-    function _borrowToAsset(uint256 amountB, uint256 clPrice) internal pure returns (uint256 assets) {
-        // The first division gets rid of the decimals of wmatic. The second converts dollars to usdc
-        assets = amountB.mulWadDown(clPrice) / 1e2;
+    /**
+     * @dev Convert `borrow` (e.g. MATIC) to `asset` (e.g. USDC)
+     * 10^borrow_decimals in `borrow` = clPrice / 10^borrowFeed_decimals * 10^asset_decimals in `assets`
+     * thus amountB in `borrow` = amountB * clPrice * 10^(asset_decimals - borrow_decimals - borrowFeed_decimals)
+     * Also note that, decimalAdjust = abs(asset_decimals - borrow_decimals - borrowFeed_decimals)
+     */
+    function _borrowToAsset(uint256 amountB, uint256 clPrice) internal view returns (uint256 assets) {
```
Comment (larryob) [inline src/both/DeltaNeutralLpV3.sol:523]: Change amountB to `borrows`

File: src/both/DeltaNeutralLpV3.sol [hunk 13]
```diff
@@ -471,15 +514,27 @@ contract DeltaNeutralLpV3 is AccessStrategy {
     /// @dev Gives ratio of vault asset to borrow asset, e.g. WMATIC/USD (we assume that usd = usdc)
     AggregatorV3Interface immutable borrowFeed;
 
-    /// @dev Convert `borrow` (e.g. MATIC) to `asset` (e.g. USDC)
-    function _borrowToAsset(uint256 amountB, uint256 clPrice) internal pure returns (uint256 assets) {
-        // The first division gets rid of the decimals of wmatic. The second converts dollars to usdc
-        assets = amountB.mulWadDown(clPrice) / 1e2;
+    /**
+     * @dev Convert `borrow` (e.g. MATIC) to `asset` (e.g. USDC)
+     * 10^borrow_decimals in `borrow` = clPrice / 10^borrowFeed_decimals * 10^asset_decimals in `assets`
```
Comment (larryob) [inline src/both/DeltaNeutralLpV3.sol:519]: This seems to be backwards. The clPrice is the asset/borrow ratio. This comment implies that

```
1 borrow / 1 asset = clPrice / borrowFeed_decimals
```

I think you can put some comments inside the function (not natspec) explaining that when we multiple `borrow` amounts by the `clPrice` we have a number with `borrowDecimals + borrowFeedDecimals`. We optionally add or remove decimals based on the value of `assetDecimals`

**Review by larryob — COMMENTED** [review]
Looking good, left more comments.

**Pull request merged**